src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java

```diff
@@ -0,0 +1,401 @@
+import org.apache.cassandra.streaming.messages.KeepAliveMessage;
+import org.apache.cassandra.streaming.messages.OutgoingFileMessage;
+import org.apache.cassandra.streaming.messages.StreamInitMessage;
+import org.apache.cassandra.streaming.messages.StreamMessage;
+import org.apache.cassandra.utils.FBUtilities;
+
+/**
+ * Responsible for sending {@link StreamMessage}s to a given peer. We manage an array of netty {@link Channel}s
+ * for sending {@link OutgoingFileMessage} instances; all other {@link StreamMessage} types are sent via
+ * the normal internode messaging via {@link MessagingService}. The reason for this is to treat those messages
+ * like any other cassandra message type, and only special case the {@link OutgoingFileMessage} as it is
+ * the one doing the heavy lifting of transferring sstables. Also, multiple netty channels allow multiple files
+ * to be sent in parallel.
+ *
+ * One of the challenges when sending files is we might need to delay shipping the file if
+ *
+ * - we've exceeded our network I/O use due to rate limiting (at the cassandra level)
+ * - the receiver isn't keeping up, which causes the local socket buffer to not empty, which causes epoll writes to not
+ * move any bytes to the socket, which causes buffers to stick around in user-land (a/k/a cassandra) memory.
+ *
+ * When those conditions occur, it's easy enough to reschedule processing the file once the resouurces pick up
+ * (we acquire the permits from the rate limiter, or the socket drains). However, we need to ensure that
+ * no other messages are submitted to the same channel while the current file is still being processed.
+ *
+ * // TODO:JEB cleanup this comment
+ * There's also a couple of tricky sizes that should be pointed out:
+ * - amount of data in memory (in netty outbound channel) - not affected by zero-copy transfers as nothing comes into user-space
+ * - amount of data going out (network bytes)
+ * - buffer sizes for reading chunks off disk - sorta not affected by zero-copy
+ */
+public class NettyStreamingMessageSender implements StreamingMessageSender
+{
+    private static final Logger logger = LoggerFactory.getLogger(NettyStreamingMessageSender.class);
+
+    /**
+     *
+     */
+    private static final int DEFAULT_CHANNEL_BUFFER_SIZE = 1 << 22;
+
+    // TODO:JEB should this be configurable? also, is this local to this stream session or global to all stream sessions or per-remote peer?
+    private static final int DEFAULT_MAX_PARALLEL_TRANSFERS = 4;
+
+    private static final long MAX_WAIT_TIME_NANOS = TimeUnit.SECONDS.toNanos(30);
+    private static final int MAX_CONNECT_ATTEMPTS = 3;
+
+    private final StreamSession session;
+    private final int protocolVersion;
+    private final OutboundConnectionIdentifier connectionId;
+    private final ServerEncryptionOptions encryptionOptions;
+
+    private final ExecutorService fileTransferExecutor;
+
+    private Channel controlMessageChannel;
```

Comment:
I'm not sure what the story is with the control channel right, but there still isn't a comment?